DIPs/DIP1xxx-RC.md

```diff
@@ -0,0 +1,232 @@
+
+### Grammar changes
+
+```diff
+TemplateParameters:
++    < NamedTemplateParameterList|opt >
+
+TemplateParameter:
++    < NamedTemplateParameterList|opt >
+
++ NamedTemplateParameterList:
++    NamedTemplateParameter
++    NamedTemplateParameter ,
++    NamedTemplateParameter , NamedTemplateParameterList
+
++ NamedTemplateParameter:
++    Identifier = TemplateParameter
++    alias Identifier
+
+TemplateArgument:
++   NamedTemplateArgumentList
+
++ NamedTemplateArgumentList:
++     NamedTemplateArgument
++     NamedTemplateArgument ,
++     NamedTemplateArgument , NamedTemplateArgumentList
+
++ NamedTemplateArgument:
++     Identifier : TemplateArgument
+
+Parameter:
++   < NamedParameterList|opt >
+
++ NamedParameterList:
++    Parameter
++    Parameter ,
++    Parameter , NamedParameterList
+
++ NamedArgument:
++    Identifier : ConditionalExpression
+
+TraitsKeyword:
++    getNamedParameters
+```
+
+### Use cases
+
+#### Ranges
+
+Removes the usage of ``ElementType`` and initiations of it by making the type used a member.
+
+```D
+struct Adder<SourceType, Type=SourceType.Type> {
+    SourceType source;
+    Type toAdd;
+
+    @property {
+        Type front() {
+            return source.front() + toAdd;
+        }
+        
+        bool empty() {
+            return source.empty();
+        }
+    }
+    
+    void popFront() {
+        source.popFront();
+    }
+}
+
+auto adder(Source)(Source source, Source.Type toAdd) {
+    return Adder!(SourceType: Source, Type: Source.Type)(source, toAdd);
+}
+```
+
+#### Logging
+
+Overridable but no longer interacts poorly with other arguments. Requiring no smelly work arounds.
+
+```D
+void log(T...)(T args, <string moduleName = __MODULE__, uint lineNumber = __LINE__>) {
+    writeln(moduleName, "[", lineNumber, "] ", args);
+}
+```
+
+## Future proposals
+
+This DIP can be expanded upon to catch more use cases (but would make a much more complex initial one).
+Here are some ideas for future DIP's.
+
+### Restrictions
+
+This DIP has a very loose definition of the parameters/arguments. In other languages they have ordering and other restrictions in place.
+While this is a point of contention, this DIP will not address them directly. Instead adding of artificial restrictions should be left to a future DIP after the community has had some experience in using them.
+
+### API Alias parameters
+
+Alias attribute to renamed parameters (templates and function ones).
+
+```D
+void func(int foo, string bar)
+alias(foo=[food, foo2], bar=offset) {}
+```
+
+Multiple attributes could be applied but you must pick one set of identifiers used inside a single attribute e.g. either "foo", "food" or "foo2" and either "bar" or "offset" at call/initaition site.
+
+Identifiers should note overlap between attributes so:
+
+```D
+void func(int foo, string bar)
+alias(foo=baz, bar=har)

```

Comment:
Suggestion, should just be `alias(baz, har)`, or `alias(foo:baz, bar:har)`, for consistency

Reply:
It could be yes. But no point changing it since this DIP doesn't introduce it, just suggest it for somebody in the future.